docs/config_reference.md
# Config Reference

This document lists all Config Options and Keywords supported by Rubberize, along with their effects.

## Config Options

Config Options are name-value pairs (e.g., `@use_symbols=True`) that control how Rubberize renders expressions and statements.

### Variable Name Rendering Options

> *Also see [variables rendering guide](rendering/variables.md).*

| Option            | Type           | Default                            | Description                                                                                                                                            |
|-------------------|----------------|------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------|
| `@use_subscripts` | `bool`         | `True`                             | Convert underscores (`_`) to subscripts.                                                                                                               |
| `@use_symbols`    | `bool`         | `True`                             | Convert greek names, accents, and letters to proper mathematical symbols.                                                                              |
| `@greek_starts`   | `set[str]`[^1] | `{"Delta", "gamma", "phi", "psi"}` | Mathematical symbols (i.e., Greek letters) that will be rendered when it appears in the beginning of the base name, e.g., `DeltaT` becomes $\Delta T$. |
| `@hidden_modules` | `set[str]`[^1] | `{"math", "sp", "np", "ureg"}`     | Names found in the set will be skipped from being rendered.                                                                                            |

### String Types Rendering Options

| Option      | Type  | Default | Description                                                                                                                                                                     |
|-------------|-------|---------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `@str_font` | `str` | `""`    | Set the $\LaTeX$ font style for `str` types. Options are `""` for default, `"bf"` for bold, `"it"` for italics, `"rm"` for roman, `"sf"` for sans serif, `"tt"` for typewriter. |

### Number Types Rendering Options

> *Also see [numbers rendering guide](rendering/numbers.mds) for options for number types.*  

| Option                   | Type    | Default | Description                                                                                                                                                                                                                                                                         |
|--------------------------|---------|---------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `@num_format`            | `str`   | `"FIX"` | Set the number format used to render real number types. Options are `"FIX"` for fixed-point decimal notation, `"SCI"` for scientific notation, `"GEN"` for general notation, `"ENG"` for engineering notation.                                                                      |
| `@num_format_prec`       | `int`   | `2`     | If `@num_format="FIX"`, set the **number of decimal digits** to be displayed. Otherwise, it is the **number of significant figures** to be displayed.                                                                                                                               |
| `@num_format_max_digits` | `int`   | `15`    | Only works when `@num_format="FIX"`. Real numbers with more digits than this value will automatically switch to scientific notation for readability.                                                                                                                                |
| `@num_format_e_not`      | `bool`  | `False` | For `@num_format` that uses the scientific notation (`"SCI"`,`"GEN"`, or `"ENG"`), use [E notation](https://en.wikipedia.org/wiki/Scientific_notation#E_notation) instead of the standard scientific notation format, e.g., $6.942\mathrm{E}{+4}$ instead of $6.942 \times 10^{4}$. |
| `@thousands_separator`   | `str`   | `" "`   | Set the thousands separator. Also applied to `int` types. Options are `""` for no separator, `" "` for thin space, `","` for comma, `"."` for point, `"'"` for apostrophe.                                                                                                          |
| `@decimal_marker`        | `str`   | `"."`   | Set the decimal marker. Options are `"."` or `","`.                                                                                                                                                                                                                                 |
| `@zero_float_threshold`  | `float` | `1e-12` | If the absolute value of the number is less than this value, it will be rendered as zero. This behavior accounts for floating-point precision errors in computations.                                                                                                               |
| `@use_polar`             | `bool`  | `False` | Display `complex` types in polar form.                                                                                                                                                                                                                                              |
| `@use_polar_deg`         | `bool`  | `True`  | Only works when `@use_polar=True`. Use degrees for phase angles of complex numbers instead of radians.                                                                                                                                                                              |

### Collection Types Rendering Options

> *Also see [collections rendering guide](rendering/collections.md) for option for collection types.*

| Option               | Type   | Default | Description                       |
|----------------------|--------|---------|-----------------------------------|
| `@show_list_as_col`  | `bool` | `True`  | Show lists as column arrays.      |
| `@show_tuple_as_col` | `bool` | `False` | Show tuples as column arrays.     |
| `@show_set_as_col`   | `bool` | `False` | Display sets on a column.         |
| `@show_dict_as_col`  | `bool` | `True`  | Display dictionaries on a column. |

### Pint Rendering Options

> *Also see [Pint rendering guide](rendering/pint.md) for options for Pint quantity and unit types.*

| Option              | Type   | Default | Description                                                                                                                                                                                                                                |
|---------------------|--------|---------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `@use_inline_units` | `bool` | `True`  | Display fractions in units using a solidus ($a/b$) instead of a fraction ($\frac{a}{b}$) for a more compact but readable appearance. If the fraction has multiple denominator terms, negative exponents will be used instead of a solidus. |
| `@use_dms_units`    | `bool` | `False` | Display angle quantities in degree-minute-second (DMS) format.                                                                                                                                                                             |
| `@use_fif_units`    | `bool` | `False` | Display length quantities in foot-inch-fraction (FIF) format.                                                                                                                                                                              |
| `@fif_prec`         | `int`  | `16`    | Only works when `@use_fif_units=True`. Set the precision of the fractional part in FIF units, e.g., `16` for 1/16 of an inch precision.                                                                                                    |

### Expression Display Mode Options

Display modes are the distinct forms an expression can take: its general definition, its substituted version with values, and its final evaluated result. Also see [expression display modes guide](rendering/expressions.md#expression-display-modes).

| Option               | Type           | Default                        | Description                                                                              |
|----------------------|----------------|--------------------------------|------------------------------------------------------------------------------------------|
| `@show_definition`   | `bool`         | `True`                         | Show the base form of the expression, before specific values have been substituted.      |
| `@show_substitution` | `bool`         | `True`                         | Show the expression after substituting numerical values to variables.                    |
| `@show_result`       | `bool`         | `True`                         | Show the final calculated value for the expression.                                      |
| `@multiline`         | `bool`         | `False`                        | Arrange display modes such that each mode is on its own line.                            |
| `@math_constants`    | `set[str]`[^1] | `{"e", "pi", "phi", "varphi"}` | Symbol names found in this set will not be substituted in the substitution display mode. |

### Expression Rendering Options

| Option                   | Type   | Default | Description                                                                                                                                                                                                          |
|--------------------------|--------|---------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `@wrap_indices`          | `bool` | `True`  | Enclose element access indices and dictionary keys in parentheses. Also see [collections guide for accessing elements](rendering/collections.md#accessing-collection-elements).                                      |
| `@convert_special_funcs` | `bool` | `True`  | Prevent the rendering of class and function calls based on defined special conversion rules.                                                                                                                         |
| `@use_contextual_mult`   | `bool` | `True`  | Use a multiplication symbol appropriate to the factors being multiplied. Also see [expressions guide for contextual multiplication](rendering/expressions.md#contextual-multiplication-symbol) for more information. |

[^1]: When passing Config Options that take sets as arguments to `%%tap`, wrap the set in quotes, e.g., `@greek_starts='{"pi","nu"}'`

## Keywords

Keywords are shorthand names that represent one or more Config Options, allowing quick adjustments.

| Keyword  | Equivalent Config                                                    | Description                                       |
|----------|----------------------------------------------------------------------|---------------------------------------------------|
| `@none`  | `@show_definition=False @show_substitution=False @show_result=False` | Hide all display modes.                           |
| `@all`   | `@show_definition=True @show_substitution=True @show_result=True`    | Show all display modes.                           |
| `@def`   | `@show_definition=True @show_substitution=False @show_result=False`  | Show only the definition display mode.            |
| `@sub`   | `@show_definition=False @show_substitution=True @show_result=False`  | Show only the substitution display mode.          |
| `@res`   | `@show_definition=False @show_substitution=False @show_result=True`  | Show only the results display mode.               |
| `@nodef` | `@show_definition=False @show_substitution=True @show_result=True`   | Show all display modes except the definition.     |
| `@nosub` | `@show_definition=True @show_substitution=False @show_result=True`   | Show all display modes except the substitution.   |
| `@nores` | `@show_definition=True @show_substitution=True @show_result=False`   | Show all display modes except the result.         |
| `@line`  | `@multiline=False`                                                   | Arrange display modes on a line.                  |
| `@stack` | `@multiline=True`                                                    | Stack the display modes.                          |
| `@fix`   | `@num_format="FIX"`                                                  | Use a fixed-point decimal notation number format. |
| `@sci`   | `@num_format="SCI"`                                                  | Use a scientific notation number format.          |
| `@gen`   | `@num_format="GEN"`                                                  | Use a general notation number format.             |
| `@eng`   | `@num_format="ENG"`                                                  | Use an engineering notation number format.        |
| `@0`     | `@num_format_prec=0`                                                 |                                                   |
| `@1`     | `@num_format_prec=1`                                                 |                                                   |
| `@2`     | `@num_format_prec=2`                                                 |                                                   |
| `@3`     | `@num_format_prec=3`                                                 |                                                   |
| `@4`     | `@num_format_prec=4`                                                 |                                                   |
| `@5`     | `@num_format_prec=5`                                                 |                                                   |
| `@6`     | `@num_format_prec=6`                                                 |                                                   |

rubberize/config.py
"""Config system.

The configuration is a singleton instance of `_Config`.
"""

import ast
import json
from contextlib import contextmanager
from dataclasses import dataclass, field, asdict
from pathlib import Path
from typing import Any, Iterable, Literal, Optional

from rubberize import exceptions


@dataclass
class _ConfigDefaults:  # pylint: disable=too-many-instance-attributes
    """Default config."""

    # Name
    use_subscripts: bool = True
    use_symbols: bool = True
    greek_starts: set[str] = field(
        default_factory=lambda: {"Delta", "gamma", "phi", "psi"}
    )

    # Display modes
    show_definition: bool = True
    show_substitution: bool = True
    show_result: bool = True

    # String display options
    str_font: Literal["", "bf", "it", "rm", "sf", "tt"] = ""

    # Number display options
    num_format: Literal["FIX", "SCI", "GEN", "ENG"] = "FIX"
    num_format_prec: int = 2
    num_format_max_digits: int = 15
    num_format_e_not: bool = False
    thousands_separator: Literal["", " ", ",", ".", "'"] = " "
    decimal_marker: Literal[".", ","] = "."
    zero_float_threshold: float = 1e-12

    use_polar: bool = False
    use_polar_deg: bool = True

    use_inline_units: bool = True
    use_dms_units: bool = False
    use_fif_units: bool = False
    fif_prec: int = 16

    # Expressions
    wrap_indices: bool = True
    convert_special_funcs: bool = True
    use_contextual_mult: bool = True
    hidden_modules: set[str] = field(
        default_factory=lambda: {"math", "sp", "np", "ureg"}
    )
    math_constants: set[str] = field(
        default_factory=lambda: {"e", "pi", "phi", "varphi"}
    )
    show_list_as_col: bool = True
    show_tuple_as_col: bool = False
    show_set_as_col: bool = False
    show_dict_as_col: bool = True

    # Statements
    multiline: bool = False


class _Config(_ConfigDefaults):
    """Singleton global configuration."""

    def __init__(self):
        super().__init__()
        self.load()

    def set(self, **kwargs: bool | int | Iterable[str]) -> None:
        """Update multiple config values passed as keyword arguments."""

        for key, value in kwargs.items():
            if not hasattr(self, key):
                raise AttributeError(f"Invalid config key: {key}")

            if key in ("greek_starts", "hidden_modules", "math_constants"):
                if not isinstance(value, (set, list, tuple)):
                    raise exceptions.RubberizeTypeError(
                        f"Invalid {key} type: {type(value)}"
                    )
                value = set(value)

            setattr(self, key, value)

    def load(self, *args: str, path: Optional[str | Path] = None) -> None:
        """Load config from defaults or a specified JSON file.

        If a file exists for a given `path`, its contents are loaded and
        used to update the config. Otherwise, default values are used.

        Args:
            *args: If provided, only the specified keys are updated.
            path: Path to the JSON file. If `None`, the default values
                are used.
        """

        data = asdict(_ConfigDefaults())

        if path is not None:
            path = Path(path)
            if path.exists():
                with path.open("r", encoding="utf-8") as f:
                    data.update(json.load(f))

        if args:
            data = {k: data[k] for k in args if k in data}
        self.set(**data)

    def reset(self, *args: str) -> None:
        """Reset the config or only the specified keys to defaults."""

        self.load(*args, path=None)

    def save(self, path: str | Path) -> None:
        """Save the current config to a JSON file."""

        config_dict = asdict(self)
        config_dict["greek_starts"] = list(config_dict["greek_starts"])
        config_dict["hidden_modules"] = list(config_dict["hidden_modules"])
        config_dict["math_constants"] = list(config_dict["math_constants"])

        path = Path(path)
        path.parent.mkdir(parents=True, exist_ok=True)

        with path.open("w", encoding="utf-8") as f:
            json.dump(config_dict, f, indent=4)

    def add_greek_start(self, *modules: str) -> None:
        """Add one or more modules to `greek_starts`."""

        self.greek_starts.update(modules)

    def remove_greek_start(self, *modules: str) -> None:
        """Remove one or more modules from `greek_starts`."""

        self.greek_starts.difference_update(modules)

    def add_hidden_module(self, *modules: str) -> None:
        """Add one or more modules to `hidden_modules`."""

        self.hidden_modules.update(modules)

    def remove_hidden_module(self, *modules: str) -> None:
        """Remove one or more modules from `hidden_modules`."""

        self.hidden_modules.difference_update(modules)

    def add_math_constant(self, *constant: str) -> None:
        """Add one or more constant to `math_constants`."""

        self.math_constants.update(constant)

    def remove_math_constant(self, *constant: str) -> None:
        """Remove one or more constant from `math_constants`."""

        self.math_constants.difference_update(constant)

    @contextmanager
    def override(self, **kwargs: bool | int | Iterable[str]):
        """Temporarily override config values within a context."""

        original_values = {key: getattr(self, key) for key in kwargs}

        try:
            self.set(**kwargs)
            yield
        finally:
            self.set(**original_values)


# fmt: off
_KEYWORDS: dict[str, dict[str, Any]] = {
    "none": {"show_definition": False, "show_substitution": False, "show_result": False},
    "all": {"show_definition": True, "show_substitution": True, "show_result": True},
    "def": {"show_definition": True, "show_substitution": False, "show_result": False},
    "sub": {"show_definition": False, "show_substitution": True, "show_result": False},
    "res": {"show_definition": False, "show_substitution": False, "show_result": True},
    "nodef": {"show_definition": False, "show_substitution": True, "show_result": True},
    "nosub": {"show_definition": True, "show_substitution": False, "show_result": True},
    "nores": {"show_definition": True, "show_substitution": True, "show_result": False},
    "line": {"multiline": False},
    "stack": {"multiline": True},
    "fix": {"num_format": "FIX"},
    "sci": {"num_format": "SCI"},
    "gen": {"num_format": "GEN"},
    "eng": {"num_format": "ENG"},
    "0": {"num_format_prec": 0},
    "1": {"num_format_prec": 1},
    "2": {"num_format_prec": 2},
    "3": {"num_format_prec": 3},
    "4": {"num_format_prec": 4},
    "5": {"num_format_prec": 5},
    "6": {"num_format_prec": 6},
}
# fmt: on


def parse_modifiers(modifiers: list[str]) -> dict[str, Any]:
    """Parse a list of modifiers to an overrides dict for use with
    `config.override()`.
    """

    override_dict = {}
    for modifier in modifiers:
        modifier = modifier.removeprefix("@")

        if modifier == "hide":
            return {"hide": ...}
        if modifier == "endhide":
            return {"endhide": ...}

        if "=" in modifier:
            key, value = modifier.split("=", 1)
            override_dict[key] = ast.literal_eval(value)
        elif modifier in _KEYWORDS:
            override_dict.update(_KEYWORDS[modifier])
        else:
            raise exceptions.RubberizeKeywordError(
                f"Unknown keyword: '{modifier}'"
            )
    return override_dict


# Singleton instance
config = _Config()
